### PR TITLE
RDBC-473 - implementation of cluster operation

### PR DESCRIPTION
**Pull request by omerlewitz**

### CODE REVIEW

**Comment by omerlewitz** [comment]
https://issues.hibernatingrhinos.com/issue/RDBC-473

**Comment by ayende** [comment]
Please squash those PRs, we don't need the "make this stop hurt me" commit messages, etc.

File: go.mod [hunk 1]
```diff
@@ -1,4 +1,6 @@
-module github.com/ravendb/ravendb-go-client
+module github.com/omerlewitz/ravendb-go-client
```
Comment (ayende) [inline go.mod:1]: No

File: operation_add_cluster_node.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package ravendb
```
Comment (ayende) [inline operation_add_cluster_node.go:1]: Let's start creating proper folder structure here.
Mimic the one we have on the JVM, please

File: operation_add_cluster_node.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	return &addNodeCommand{
+		op: operation,
+	}, nil
+}
+
+type addNodeCommand struct {
+	RavenCommandBase
+	op *OperationAddClusterNode
+}
+
```
Comment (ayende) [inline operation_add_cluster_node.go:27]: You are missing the `setResult()` here, no?

File: operation_boostrap.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+func (o* OperationBootstrap) GetCommand(conventions *DocumentConventions) (RavenCommand, error) {
+	return &bootstrapCommand{}, nil
+}
+
+type bootstrapCommand struct{
+	RavenCommandBase
+
+}
+
```
Comment (ayende) [inline operation_boostrap.go:18]: Where is `setResult()` ?

File: operation_demote_cluster_node.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	return &demoteNodeCommand{
+		op: operation,
+	}, nil
+}
+
+type demoteNodeCommand struct {
+	RaftCommandBase
+	op *OperationDemoteClusterNode
+}
+
```
Comment (ayende) [inline operation_demote_cluster_node.go:21]: again, and basically in all of them

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by ml054** [comment]
do we want "make this stop hurt me" change at all? (see https://github.com/ravendb/ravendb-go-client/pull/170/commits/3b8b4e958641b681b6b630d6010a90806afb4734)

Looks it will cause troubles after merge

**Comment by omerlewitz** [comment]
@ml054 of course we don't want it.

File: go.mod (hunk 1 above)
Comment (omerlewitz) [reply go.mod]: My remote was a mess.
I reset hard to raven-go-client and the only change I see atm is `	github.com/google/uuid v1.3.0` due to `raft id`
at the bottom of the PR page, I see that there's no mention my remote.

File: operation_add_cluster_node.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	Tag     string
+	Watcher bool
+}
+
+func (operation *OperationAddClusterNode) GetCommand(conventions *DocumentConventions) (RavenCommand, error) {
+	return &addNodeCommand{
+		op: operation,
+	}, nil
+}
+
```
Comment (ayende) [inline operation_add_cluster_node.go:22]: `SetResult` ? Move to operations folder

File: operation_boostrap.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+	"net/http"
+)
+
+type OperationBootstrap struct {
+}
+
+func (o* OperationBootstrap) GetCommand(conventions *DocumentConventions) (RavenCommand, error) {
+	return &bootstrapCommand{}, nil
+}
+
```
Comment (ayende) [inline operation_boostrap.go:13]: SetResult ? Move to operations folder

File: operation_demote_cluster_node.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+type OperationDemoteClusterNode struct {
+	Node string
+}
+
+func (operation *OperationDemoteClusterNode) GetCommand(conventions *DocumentConventions) (RavenCommand, error) {
+	return &demoteNodeCommand{
+		op: operation,
+	}, nil
+}
+
```
Comment (ayende) [inline operation_demote_cluster_node.go:16]: SetResult ? Move to operations folder

File: operation_promote_cluster_node.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	return &promoteNodeCommand{
+		op: operation,
+	}, nil
+}
+
+type promoteNodeCommand struct {
+	RaftCommandBase
+	op *OperationPromoteClusterNode
+}
+
```
Comment (ayende) [inline operation_promote_cluster_node.go:26]: SetResult ? Move to operations folder

File: operation_remove_cluster_node.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+	return &removeNodeCommand{
+		op: operation,
+	}, nil
+}
+
+type removeNodeCommand struct {
+	RavenCommandBase
+	op *RemoveClusterNode
+}
+
```
Comment (ayende) [inline operation_remove_cluster_node.go:26]: SetResult ? Move to operations folder

File: serverwide/operations/operation_add_cluster_node.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+func (c *addNodeCommand) createRequest(node *ravendb.ServerNode) (*http.Request, error) {
+	url := node.URL + "/admin/cluster/node?url=" + url.QueryEscape(c.parent.Url) + "&watcher=" + strconv.FormatBool(c.parent.Watcher)
+
+	if len(strings.TrimSpace(c.parent.Tag)) == 0 {
+		url += fmt.Sprintf("&tag=%s", c.parent.Tag)
+	}
+	return http.NewRequest(http.MethodPut, url, nil)
+}
+
```
Comment (ayende) [inline serverwide/operations/operation_add_cluster_node.go:41]: Move to operations folder

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by omerlewitz** [comment]
https://github.com/ravendb/ravendb-go-client/pull/171

**Pull request closed without merging**